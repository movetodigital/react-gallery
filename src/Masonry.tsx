import * as React from 'react';
import styled from 'styled-components';

type FunctionName = (n: object) => any;

interface ImageItemProps {
  src?: string;
  title?: string;
  alt?: string;
}

interface ComponentProps {
  images: Array<ImageItemProps>;
  gridColumns: number;
  gridGutter: number;
  imageRenderer?: FunctionName;
}

interface ImagesComponentProps {
  gridGutter: number;
  gridColumns: number;
}

interface StyleItemProps {
  gridGutter: number;
}

const Images = styled.div<ImagesComponentProps>`
  column-count: ${props => props.gridColumns};
  margin: 0 auto;
  column-gap: ${props => props.gridGutter}%;
`;

const Item = styled.div<StyleItemProps>`
  margin: 0 0 ${props => props.gridGutter}em;
`;

const DefaultImage = styled.img`
  max-width: 100%;
  max-height: 100%;
`;

interface RendererProps {
  props: ImageItemProps;
  key: number;
}

const defaultRender = ({props: {src, title, alt}, key}: RendererProps) => {
  return <DefaultImage key={src} src={src} title={title} alt={alt} />;
};

const Masonry: React.FunctionComponent<ComponentProps> = ({
  images,
  gridColumns,
  gridGutter,
  imageRenderer = defaultRender,
}) => {
  return (
    <Images gridColumns={gridColumns} gridGutter={gridGutter}>
      {images.map((el, index) => (
        <Item key={`${index}_${gridGutter}`} gridGutter={gridGutter}>
          <div>{imageRenderer({props: el, key: index})}</div>
        </Item>
      ))}
    </Images>
  );
};

export {Masonry};
