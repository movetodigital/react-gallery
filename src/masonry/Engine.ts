/* eslint-disable @typescript-eslint/no-explicit-any,@typescript-eslint/array-type */
// eslint-disable @typescript-eslint/no-explicit-any
export const COLUMNS_MAX_COUNT = 5;
export const COLUMN_MAX_WIDTH = 200;
export const GUTTER_IN_PERCENT = 0.5;
export const PLACEHOLDER_COLOR = '#f0f0f0';

const resizeByWidth = (
  item: {width: number; height: number},
  newWidth: number
) => {
  const aspectRatio = item.width / item.height;
  return {
    ...item,
    width: newWidth,
    height: newWidth / aspectRatio,
  };
};

class Engine {
  private images: Array<any>;
  private gutterInPercent: number;
  private maxColumnsCount: number;
  private columnMaxWidth: number;

  public setImages(images: Array<any>) {
    if (!images) {
      images = [];
    }
    this.images = images;
    return this;
  }

  public setGutterInPercent(gutterInPercent: number) {
    if (!gutterInPercent || gutterInPercent < 0) {
      gutterInPercent = 0;
    }
    this.gutterInPercent = gutterInPercent;
    return this;
  }

  public getGutterInPercent() {
    return this.gutterInPercent;
  }

  public setMaxColumnsCount(maxColumnsCount: number) {
    if (!maxColumnsCount || maxColumnsCount < 0) {
      maxColumnsCount = COLUMNS_MAX_COUNT;
    }
    this.maxColumnsCount = maxColumnsCount;
    return this;
  }

  public getMaxColumnsCount() {
    return this.maxColumnsCount;
  }

  public setColumnMaxWidth(columnMaxWidth: number) {
    if (!columnMaxWidth || columnMaxWidth < 0) {
      columnMaxWidth = COLUMN_MAX_WIDTH;
    }
    this.columnMaxWidth = columnMaxWidth;
    return this;
  }

  public getColumnsMaxWidth() {
    return this.columnMaxWidth;
  }

  public normalizeByWidth() {
    const calculateWidth = (item: {width: any; height?: number}) => {
      return item.width > this.columnMaxWidth
        ? this.columnMaxWidth
        : item.width;
    };
    const calculateHeight = (item: {width: number; height: number}) => {
      const itemAfterResize = resizeByWidth(item, calculateWidth(item));
      return itemAfterResize.height;
    };
    const result: Array<any> = [];
    this.images.forEach(el => {
      result.push({
        ...el,
        height: calculateHeight(el),
        width: calculateWidth(el),
      });
    });
    return result;
  }

  public buildColumns() {
    const columns = [];
    let order;
    const items = this.normalizeByWidth() as Array<any>;
    for (let i = 0; i < this.maxColumnsCount; i++) {
      columns.push({images: [], order: i});
    }
    for (let i = 0; i < items.length; i++) {
      order =
        (i + 1) % this.maxColumnsCount === 0
          ? this.maxColumnsCount
          : (i + 1) % this.maxColumnsCount;
      columns[order - 1].images.push(items[i] as never);
      items[i].order = order;
    }
    return columns;
  }
}

export default Engine;
