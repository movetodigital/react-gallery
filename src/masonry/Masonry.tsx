/* eslint-disable @typescript-eslint/no-explicit-any */
// @ts-nocheck
import React, {useEffect, useRef, useState} from 'react';

import {Container} from '../common/nodes';
import {defaultDetailsViewRenderer, defaultRenderer} from '../common/Renderers';
import ViewMonitor from '../common/ViewMonitor';
import Engine, {
  COLUMN_MAX_WIDTH,
  COLUMNS_MAX_COUNT,
  GUTTER_IN_PERCENT,
  PLACEHOLDER_COLOR,
} from './Engine';

type FunctionName = (n: object) => any;

interface ImageItemProps {
  src?: string;
  title?: string;
  alt?: string;
}

interface ComponentProps {
  imageRenderer?: FunctionName;
  images?: Array<ImageItemProps>;
  columnsMaxCount?: number;
  columnMaxWidth?: number;
  gutterInPercent?: number;
  className?: string;
  columnClassName?: string;
  enableDetailView?: boolean;
  detailsViewRenderer?: FunctionName;
  placeholderColor?: string;
}

const Masonry = ({
  imageRenderer = defaultRenderer,
  images = [],
  columnsMaxCount = COLUMNS_MAX_COUNT,
  columnMaxWidth = COLUMN_MAX_WIDTH,
  gutterInPercent = GUTTER_IN_PERCENT,
  className = '',
  columnClassName = '',
  enableDetailView = false,
  detailsViewRenderer = defaultDetailsViewRenderer,
  placeholderColor = PLACEHOLDER_COLOR,
}: ComponentProps) => {
  const engine = useRef();

  const [selectedImageRow, setSelectedImageRow] = useState(null);
  const [selectedImage, setSelectedImage] = useState(null);
  const [selectedRowHeight, setSelectedRowHeight] = useState(0);
  const [selectedImageId, setSelectedImageId] = useState(null);
  const [selectedImageProps, setSelectedImageProps] = useState({});
  const [columns, setColumns] = useState([]);

  useEffect(() => {
    engine.current = new Engine();
    engine.current
      .setImages(images)
      .setMaxColumnsCount(columnsMaxCount)
      .setColumnMaxWidth(columnMaxWidth)
      .setGutterInPercent(gutterInPercent);
    setColumns(engine.current.buildColumns());
  }, []);

  useEffect(() => {
    engine.current
      .setImages(images)
      .setMaxColumnsCount(columnsMaxCount)
      .setColumnMaxWidth(columnMaxWidth)
      .setGutterInPercent(gutterInPercent);
    setColumns(engine.current.buildColumns());
  }, [images, columnsMaxCount, columnMaxWidth, gutterInPercent]);

  const handleSelectImage = ({
    imageRow,
    imageId,
    rowHeight,
    image,
    imageProps,
  }: any) => {
    if (selectedImageRow !== imageRow || imageId !== selectedImageId) {
      setSelectedImageRow(imageRow);
      setSelectedImage(image);
      setSelectedRowHeight(rowHeight);
      setSelectedImageId(imageId);
      setSelectedImageProps(imageProps);
    } else {
      setSelectedImageRow(null);
      setSelectedImage(null);
      setSelectedRowHeight(0);
      setSelectedImageId(null);
      setSelectedImageProps({});
    }
  };

  const getColumnWidth = (gutter, count) =>
    (100 - gutter * (count - 1)) / count;

  return (
    <Container className={className}>
      {columns.map((item, columnIndex) => (
        <React.Fragment key={`image-column-${columnIndex}`}>
          <div
            /* eslint-disable-next-line react/no-array-index-key */
            key={`column-${columnIndex}`}
            className={columnClassName}
            style={{
              verticalAlign: 'top',
              position: 'relative',
              display: 'inline-block',
              width: `${getColumnWidth(
                engine.current.getGutterInPercent(),
                engine.current.getMaxColumnsCount()
              )}%`,
              margin: `0 ${
                columnIndex === columns.length - 1
                  ? 0
                  : engine.current.getGutterInPercent()
              }% 0 0`,
            }}
          >
            {item.images.map(
              (
                image: {height?: number; width?: number; src?: string},
                imageIndex: number
              ) => {
                const placeholderHeight =
                  (100 * (image.height || 0)) / (image.width || 1);
                const imageId = `column-${image.src}-${columnIndex}-${imageIndex}`;
                return (
                  <div
                    /* eslint-disable-next-line react/no-array-index-key */
                    key={`image-${image.src}-${columnIndex}-${imageIndex}`}
                    style={{
                      margin: `0 0 ${engine.current.getGutterInPercent() *
                        engine.current.getMaxColumnsCount()}% 0`,
                    }}
                  >
                    <ViewMonitor>
                      {(isViewable: boolean) =>
                        imageRenderer({
                          ...image,
                          placeholderHeight,
                          visible: isViewable,
                          enableMasonry: true,
                          placeholderColor,
                          onClick: () =>
                            handleSelectImage({
                              imageRow: columnIndex,
                              image: image,
                              rowHeight: placeholderHeight,
                              // eslint-disable-next-line
                              imageId,
                              imageProps: {
                                placeholderHeight,
                              },
                            }),
                        })
                      }
                    </ViewMonitor>
                    {enableDetailView &&
                      columnIndex === selectedImageRow &&
                      selectedImageId === imageId && (
                        <div
                          key={`detail-${imageId}`}
                          style={{
                            margin: `${engine.current.getGutterInPercent() *
                              engine.current.getMaxColumnsCount()}% 0 0 0`,
                          }}
                        >
                          {detailsViewRenderer({
                            ...item,
                            visible: columnIndex === selectedImageRow,
                            selectedImage,
                            rowHeight: selectedRowHeight,
                            gutterInPercent: engine.current.getGutterInPercent(),
                            selectedImageId,
                            selectedImageProps,
                          })}
                        </div>
                      )}
                  </div>
                );
              }
            )}
          </div>
        </React.Fragment>
      ))}
    </Container>
  );
};

export {Masonry};
