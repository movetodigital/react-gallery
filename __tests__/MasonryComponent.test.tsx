import {mount} from 'enzyme';
import * as React from 'react';

import {Masonry} from '../src/masonry/Masonry';

const hydratedProps = {
  images: [
    {
      src: '/image/name/1',
      width: 500,
      height: 500,
    },
    {
      src: '/image/name/1',
      width: 500,
      height: 500,
    },
    {
      src: '/image/name/1',
      width: 500,
      height: 500,
    },
    {
      src: '/image/name/1',
      width: 500,
      height: 500,
    },
  ],
  gridColumns: 5,
  gridGutter: 1,
};

describe('<Masonry />', () => {
  it('should render and match snapshot', () => {
    const tree = mount(<Masonry {...hydratedProps} />);

    expect(tree).toMatchSnapshot();
  });
});
