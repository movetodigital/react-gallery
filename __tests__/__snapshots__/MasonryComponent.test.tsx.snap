// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Masonry /> should render and match snapshot 1`] = `
<Masonry
  gridColumns={5}
  gridGutter={1}
  images={
    Array [
      Object {
        "height": 500,
        "src": "/image/name/1",
        "width": 500,
      },
      Object {
        "height": 500,
        "src": "/image/name/1",
        "width": 500,
      },
      Object {
        "height": 500,
        "src": "/image/name/1",
        "width": 500,
      },
      Object {
        "height": 500,
        "src": "/image/name/1",
        "width": 500,
      },
    ]
  }
>
  <Styled(Component)
    className=""
  >
    <StyledComponent
      className=""
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-htpNat",
            "isStatic": false,
            "lastClassName": "cihfGY",
            "rules": Array [
              "
  position: relative;
  display: block;
  font-size: 0;
  opacity: ",
              [Function],
              ";
  height: initial;
  overflow: initial;
",
            ],
          },
          "displayName": "Styled(Component)",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-htpNat",
          "target": Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          },
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
    >
      <ForwardRef
        className="sc-htpNat cihfGY"
      >
        <div
          className="sc-htpNat cihfGY"
        >
          <div
            className=""
            key="column-0"
            style={
              Object {
                "display": "inline-block",
                "margin": "0 0.5% 0 0",
                "position": "relative",
                "verticalAlign": "top",
                "width": "19.5%",
              }
            }
          >
            <div
              key="image-/image/name/1-0-0"
              style={
                Object {
                  "margin": "0 0 2.5% 0",
                }
              }
            >
              <ViewMonitor>
                <div>
                  <Image
                    alt=""
                    enableMasonry={true}
                    height={200}
                    onClick={[Function]}
                    order={1}
                    placeholderColor="#f0f0f0"
                    placeholderHeight={100}
                    src="/image/name/1"
                    visible={true}
                    width={200}
                  >
                    <Styled(Component)
                      alt=""
                      onClick={[Function]}
                      src="/image/name/1"
                      visible={true}
                    >
                      <StyledComponent
                        alt=""
                        forwardedComponent={
                          Object {
                            "$$typeof": Symbol(react.forward_ref),
                            "attrs": Array [],
                            "componentStyle": ComponentStyle {
                              "componentId": "sc-bwzfXH",
                              "isStatic": false,
                              "lastClassName": "eYmfNB",
                              "rules": Array [
                                "
  position: absolute;
  display: ",
                                [Function],
                                ";
  width: 100%;
  cursor: pointer;
",
                              ],
                            },
                            "displayName": "Styled(Component)",
                            "foldedComponentIds": Array [],
                            "render": [Function],
                            "styledComponentId": "sc-bwzfXH",
                            "target": [Function],
                            "toString": [Function],
                            "warnTooManyClasses": [Function],
                            "withComponent": [Function],
                          }
                        }
                        forwardedRef={null}
                        onClick={[Function]}
                        src="/image/name/1"
                        visible={true}
                      >
                        <Component
                          alt=""
                          className="sc-bwzfXH eYmfNB"
                          onClick={[Function]}
                          src="/image/name/1"
                          visible={true}
                        >
                          <img
                            alt=""
                            className="sc-bwzfXH eYmfNB"
                            onClick={[Function]}
                            src="/image/name/1"
                          />
                        </Component>
                      </StyledComponent>
                    </Styled(Component)>
                  </Image>
                  <div
                    style={
                      Object {
                        "backgroundColor": "#f0f0f0",
                        "paddingTop": "100%",
                      }
                    }
                  />
                </div>
              </ViewMonitor>
            </div>
          </div>
          <div
            className=""
            key="column-1"
            style={
              Object {
                "display": "inline-block",
                "margin": "0 0.5% 0 0",
                "position": "relative",
                "verticalAlign": "top",
                "width": "19.5%",
              }
            }
          >
            <div
              key="image-/image/name/1-1-0"
              style={
                Object {
                  "margin": "0 0 2.5% 0",
                }
              }
            >
              <ViewMonitor>
                <div>
                  <Image
                    alt=""
                    enableMasonry={true}
                    height={200}
                    onClick={[Function]}
                    order={2}
                    placeholderColor="#f0f0f0"
                    placeholderHeight={100}
                    src="/image/name/1"
                    visible={true}
                    width={200}
                  >
                    <Styled(Component)
                      alt=""
                      onClick={[Function]}
                      src="/image/name/1"
                      visible={true}
                    >
                      <StyledComponent
                        alt=""
                        forwardedComponent={
                          Object {
                            "$$typeof": Symbol(react.forward_ref),
                            "attrs": Array [],
                            "componentStyle": ComponentStyle {
                              "componentId": "sc-bwzfXH",
                              "isStatic": false,
                              "lastClassName": "eYmfNB",
                              "rules": Array [
                                "
  position: absolute;
  display: ",
                                [Function],
                                ";
  width: 100%;
  cursor: pointer;
",
                              ],
                            },
                            "displayName": "Styled(Component)",
                            "foldedComponentIds": Array [],
                            "render": [Function],
                            "styledComponentId": "sc-bwzfXH",
                            "target": [Function],
                            "toString": [Function],
                            "warnTooManyClasses": [Function],
                            "withComponent": [Function],
                          }
                        }
                        forwardedRef={null}
                        onClick={[Function]}
                        src="/image/name/1"
                        visible={true}
                      >
                        <Component
                          alt=""
                          className="sc-bwzfXH eYmfNB"
                          onClick={[Function]}
                          src="/image/name/1"
                          visible={true}
                        >
                          <img
                            alt=""
                            className="sc-bwzfXH eYmfNB"
                            onClick={[Function]}
                            src="/image/name/1"
                          />
                        </Component>
                      </StyledComponent>
                    </Styled(Component)>
                  </Image>
                  <div
                    style={
                      Object {
                        "backgroundColor": "#f0f0f0",
                        "paddingTop": "100%",
                      }
                    }
                  />
                </div>
              </ViewMonitor>
            </div>
          </div>
          <div
            className=""
            key="column-2"
            style={
              Object {
                "display": "inline-block",
                "margin": "0 0.5% 0 0",
                "position": "relative",
                "verticalAlign": "top",
                "width": "19.5%",
              }
            }
          >
            <div
              key="image-/image/name/1-2-0"
              style={
                Object {
                  "margin": "0 0 2.5% 0",
                }
              }
            >
              <ViewMonitor>
                <div>
                  <Image
                    alt=""
                    enableMasonry={true}
                    height={200}
                    onClick={[Function]}
                    order={3}
                    placeholderColor="#f0f0f0"
                    placeholderHeight={100}
                    src="/image/name/1"
                    visible={true}
                    width={200}
                  >
                    <Styled(Component)
                      alt=""
                      onClick={[Function]}
                      src="/image/name/1"
                      visible={true}
                    >
                      <StyledComponent
                        alt=""
                        forwardedComponent={
                          Object {
                            "$$typeof": Symbol(react.forward_ref),
                            "attrs": Array [],
                            "componentStyle": ComponentStyle {
                              "componentId": "sc-bwzfXH",
                              "isStatic": false,
                              "lastClassName": "eYmfNB",
                              "rules": Array [
                                "
  position: absolute;
  display: ",
                                [Function],
                                ";
  width: 100%;
  cursor: pointer;
",
                              ],
                            },
                            "displayName": "Styled(Component)",
                            "foldedComponentIds": Array [],
                            "render": [Function],
                            "styledComponentId": "sc-bwzfXH",
                            "target": [Function],
                            "toString": [Function],
                            "warnTooManyClasses": [Function],
                            "withComponent": [Function],
                          }
                        }
                        forwardedRef={null}
                        onClick={[Function]}
                        src="/image/name/1"
                        visible={true}
                      >
                        <Component
                          alt=""
                          className="sc-bwzfXH eYmfNB"
                          onClick={[Function]}
                          src="/image/name/1"
                          visible={true}
                        >
                          <img
                            alt=""
                            className="sc-bwzfXH eYmfNB"
                            onClick={[Function]}
                            src="/image/name/1"
                          />
                        </Component>
                      </StyledComponent>
                    </Styled(Component)>
                  </Image>
                  <div
                    style={
                      Object {
                        "backgroundColor": "#f0f0f0",
                        "paddingTop": "100%",
                      }
                    }
                  />
                </div>
              </ViewMonitor>
            </div>
          </div>
          <div
            className=""
            key="column-3"
            style={
              Object {
                "display": "inline-block",
                "margin": "0 0.5% 0 0",
                "position": "relative",
                "verticalAlign": "top",
                "width": "19.5%",
              }
            }
          >
            <div
              key="image-/image/name/1-3-0"
              style={
                Object {
                  "margin": "0 0 2.5% 0",
                }
              }
            >
              <ViewMonitor>
                <div>
                  <Image
                    alt=""
                    enableMasonry={true}
                    height={200}
                    onClick={[Function]}
                    order={4}
                    placeholderColor="#f0f0f0"
                    placeholderHeight={100}
                    src="/image/name/1"
                    visible={true}
                    width={200}
                  >
                    <Styled(Component)
                      alt=""
                      onClick={[Function]}
                      src="/image/name/1"
                      visible={true}
                    >
                      <StyledComponent
                        alt=""
                        forwardedComponent={
                          Object {
                            "$$typeof": Symbol(react.forward_ref),
                            "attrs": Array [],
                            "componentStyle": ComponentStyle {
                              "componentId": "sc-bwzfXH",
                              "isStatic": false,
                              "lastClassName": "eYmfNB",
                              "rules": Array [
                                "
  position: absolute;
  display: ",
                                [Function],
                                ";
  width: 100%;
  cursor: pointer;
",
                              ],
                            },
                            "displayName": "Styled(Component)",
                            "foldedComponentIds": Array [],
                            "render": [Function],
                            "styledComponentId": "sc-bwzfXH",
                            "target": [Function],
                            "toString": [Function],
                            "warnTooManyClasses": [Function],
                            "withComponent": [Function],
                          }
                        }
                        forwardedRef={null}
                        onClick={[Function]}
                        src="/image/name/1"
                        visible={true}
                      >
                        <Component
                          alt=""
                          className="sc-bwzfXH eYmfNB"
                          onClick={[Function]}
                          src="/image/name/1"
                          visible={true}
                        >
                          <img
                            alt=""
                            className="sc-bwzfXH eYmfNB"
                            onClick={[Function]}
                            src="/image/name/1"
                          />
                        </Component>
                      </StyledComponent>
                    </Styled(Component)>
                  </Image>
                  <div
                    style={
                      Object {
                        "backgroundColor": "#f0f0f0",
                        "paddingTop": "100%",
                      }
                    }
                  />
                </div>
              </ViewMonitor>
            </div>
          </div>
          <div
            className=""
            key="column-4"
            style={
              Object {
                "display": "inline-block",
                "margin": "0 0.5% 0 0",
                "position": "relative",
                "verticalAlign": "top",
                "width": "19.5%",
              }
            }
          />
        </div>
      </ForwardRef>
    </StyledComponent>
  </Styled(Component)>
</Masonry>
`;
